### PR TITLE
[LibraryOfCongress] Added new extractor (Closes #3188)

### LINKED ISSUE #3118: Problem whit software
I can not dowload anymore 

WARNING: Warning: Falling back to static signature algorithm
[youtube] kigOAUK0hCQ: Downloading js player vflazPBMd
WARNING: Automatic signature extraction failed: Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/youtube_dl/extractor/youtube.py", line 874, in _decrypt_signature
    video_id, player_url, len(s)
  File "/usr/lib/python2.7/dist-packages/youtube_dl/extractor/youtube.py", line 376, in _extract_signature_function
    res = self._parse_sig_js(code)
  File "/usr/lib/python2.7/dist-packages/youtube_dl/extractor/youtube.py", line 444, in _parse_sig_js
    u'Initial JS player signature function name')
  File "/usr/lib/python2.7/dist-packages/youtube_dl/extractor/common.py", line 383, in _search_regex
    raise RegexNotFoundError(u'Unable to extract %s' % _name)
RegexNotFoundError: Unable to extract Initial JS player signature function name; please report this issue on https://yt-dl.org/bug . Be sure to call youtube-dl with the --verbose flag and include its complete output. Make sure you are using the latest version; type  youtube-dl -U  to update.

WARNING: Warning: Falling back to static signature algorithm
[youtube] kigOAUK0hCQ: Downloading js player vflazPBMd

### LINKED ISSUE #3188: Download from Library of Congress
I'm trying to get http://www.loc.gov/item/90716351/ , for which there seems to be no option to download. I've tried that and 

[debug] System config: []
[debug] User config: []
[debug] Command-line args: ['--verbose', 'http://www.loc.gov/item/90716351/']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2014.06.26
[debug] Python version 2.7.7 - Linux-3.14-1-amd64-x86_64-with-debian-jessie-sid
[debug] Proxy map: {}
[generic] 90716351: Requesting header
WARNING: Falling back on generic information extractor.
[generic] 90716351: Downloading webpage
[generic] 90716351: Extracting information
ERROR: Unsupported URL: http://www.loc.gov/item/90716351/; please report this issue on https://yt-dl.org/bug . Be sure to call youtube-dl with the --verbose flag and include its complete output. Make sure you are using the latest version; type  youtube-dl -U  to update.
Traceback (most recent call last):
  File "/home2/prosfilaes/bin/bin/youtube-dl/youtube_dl/extractor/generic.py", line 451, in _real_extract
    doc = parse_xml(webpage)
  File "/home2/prosfilaes/bin/bin/youtube-dl/youtube_dl/utils.py", line 1406, in parse_xml
    return xml.etree.ElementTree.XML(s.encode('utf-8'), **kwargs)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 1300, in XML
    parser.feed(text)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 1642, in feed
    self._raiseerror(v)
  File "/usr/lib/python2.7/xml/etree/ElementTree.py", line 1506, in _raiseerror
    raise err
ParseError: mismatched tag: line 331, column 2
Traceback (most recent call last):
  File "/home2/prosfilaes/bin/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 516, in extract_info
    ie_result = ie.extract(url)
  File "/home2/prosfilaes/bin/bin/youtube-dl/youtube_dl/extractor/common.py", line 167, in extract
    return self._real_extract(url)
  File "/home2/prosfilaes/bin/bin/youtube-dl/youtube_dl/extractor/generic.py", line 741, in _real_extract
    raise ExtractorError('Unsupported URL: %s' % url)
ExtractorError: Unsupported URL: http://www.loc.gov/item/90716351/; please report this issue on https://yt-dl.org/bug . Be sure to call youtube-dl with the --verbose flag and include its complete output. Make sure you are using the latest version; type  youtube-dl -U  to update.

The following seems to be the chunk of code that implements the video.
    <div class='largeview mediaobject
     videoobject '>
        <script type="text/javascript"
                src='http://media.loc.gov/loader/js/media.js'/>
        </script>
		
            <div id='mediaInstance1' class='locmedia-load-this' style='display:none'>
                <code>{
    "mimetype": "application/x-video", 
    "url": "http://lccn.loc.gov/90716351#cosmos_set", 
    "mediaType": "V", 
    "detailUrl": "http://lccn.loc.gov/90716351#cosmos_set", 
    "mediaObjectId": "E6AB0B2585930180E0438C93F0280180", 
    "aspectRatio": "4:3", 
    "autoPlay": false, 
    "shortName": "Pa's trip to Mars", 
    "playerSize": "largeStandard", 
    "metadata": [
        "Pa's trip to Mars"
    ]
}</code> 
            </div>

### PR DESCRIPTION
**Pull request by TRox1972**
Added extractor of loc.gov, which closes #3188. I am not an experienced programmer, so I am sure I did a bunch of mistakes, but the extractor works (for me at least).

### CODE REVIEW

**Comment by TRox1972** [comment]
I know I mistakenly added multiple independent commits in one pull request. I'll close my other PR and avoid it next time. Apologies.

**Comment by TRox1972** [comment]
Note that the extractor does not include urls in the from of https://www.loc.gov/item/mp<id>, as they have can easily be downloaded from a browser (download option) and the generic extractor also seems to be able to download them.

File: Makefile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 all: youtube-dl README.md CONTRIBUTING.md README.txt youtube-dl.1 youtube-dl.bash-completion youtube-dl.zsh youtube-dl.fish supportedsites
 
 clean:
-	rm -rf youtube-dl.1.temp.md youtube-dl.1 youtube-dl.bash-completion README.txt MANIFEST build/ dist/ .coverage cover/ youtube-dl.tar.gz youtube-dl.zsh youtube-dl.fish youtube_dl/extractor/lazy_extractors.py *.dump *.part *.info.json *.mp4 *.flv *.mp3 *.avi *.mkv *.webm CONTRIBUTING.md.tmp ISSUE_TEMPLATE.md.tmp youtube-dl youtube-dl.exe
+	rm -rf youtube-dl.1.temp.md youtube-dl.1 youtube-dl.bash-completion README.txt MANIFEST build/ dist/ .coverage cover/ youtube-dl.tar.gz youtube-dl.zsh youtube-dl.fish youtube_dl/extractor/lazy_extractors.py *.dump *.part *.info.json *.mp4 *.flv *.mp3 *.avi *.mkv *.webm *.jpg *.png CONTRIBUTING.md.tmp ISSUE_TEMPLATE.md.tmp youtube-dl youtube-dl.exe
```
Comment (dstftw) [inline Makefile]: This change is unrelated.

File: youtube_dl/extractor/libraryofcongress.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# coding: utf-8
+
+from __future__ import unicode_literals
+
+import re
+
+from .common import InfoExtractor
+
```
Comment (dstftw) [inline youtube_dl/extractor/libraryofcongress.py]: Check code with flake8 and fix all issues.

File: youtube_dl/extractor/libraryofcongress.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
+        
+        m3u8_id = re.search(r'media_service [0-9]{7}-[0-9]-[0-9]', webpage).group()[14:]
+        m3u8_url = 'http://stream-media.loc.gov/hls-vod/media/ls/sagan/%s.mp4.m3u8' % m3u8_id
```
Comment (dstftw) [inline youtube_dl/extractor/libraryofcongress.py]: Correct approach is to mimic web browser:
1. Use https://media.loc.gov/services/v1/media?callback=jQuery18205494330186516222_1463503121265&id=E6AB0B2585930180E0438C93F0280180&context=jsonp&_=1463503124082
2. Mimic web player url construction:
```js
        // generate hls live streaming url over https
        url = url.replace("rtmp", "http");
        
        // some urls come in with no extension and we need it on there for access
        if(url.indexOf('.mp4') == -1 && url.indexOf('.mp3') == -1){
            url += (is_video ? '.mp4' : '.mp3');
        }
        
        if(url.indexOf('vod/mp4:') > -1 ){
            url = url.replace("vod/mp4:", "hls-vod/media/")+".m3u8";
        } else if ( url.indexOf('vod/mp3:') > -1 ){
            url = url.replace("vod/mp3:", "");
        }
```

**Comment by TRox1972** [comment]
thanks for feedback @dstftw. Does this seem good?

File: youtube_dl/extractor/libraryofcongress.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        self.report_download_webpage(video_id)
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
+        json_id = re.search('media-player-[0-9A-Z]{32}', webpage).group(0).replace('media-player-', '')
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: Use ```self._search_regex```.

File: youtube_dl/extractor/libraryofcongress.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            media_url = media_url.replace('vod/mp4:', 'hls-vod/media/') + '.m3u8'
+        elif url.index('vod/mp3:') > -1:
+            media_url = media_url.replace('vod/mp3:', '')
+
+        if media_url.endswith('.m3u8'):
+            formats = self._extract_m3u8_formats(media_url, video_id, ext='mp4')
+        elif media_url.endswith('.mp3'):
+            formats = media_url
+
+        thumbnail_url = re.search(r'http://stream-media.loc.gov/ls/sagan/stills/[0-9]{7}-[0-9]-[0-9]_thumb.jpg',
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: Use ```self._search_regex``` or ```self._html_search_regex```. And an JPEG URL only does not indicate it's the thumbnail. A stricter filter needed.

File: youtube_dl/extractor/libraryofcongress.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        self.report_download_webpage(video_id)
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
+        json_id = re.search('media-player-[0-9A-Z]{32}', webpage).group(0).replace('media-player-', '')
+        data = self._parse_json(self._download_webpage('https://media.loc.gov/services/v1/media?id=%s' %json_id, video_id), video_id)
+        data = data.get('mediaObject')
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: ```.get()``` idiom is used for optional fields only.

File: youtube_dl/extractor/libraryofcongress.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            media_url += '.mp4' if is_video else '.mp3'
+
+        if media_url.index('vod/mp4:') > -1:
+            media_url = media_url.replace('vod/mp4:', 'hls-vod/media/') + '.m3u8'
+        elif url.index('vod/mp3:') > -1:
+            media_url = media_url.replace('vod/mp3:', '')
+
+        if media_url.endswith('.m3u8'):
+            formats = self._extract_m3u8_formats(media_url, video_id, ext='mp4')
+        elif media_url.endswith('.mp3'):
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: Better to use ```determine_ext``` instead of ```.endswith```

File: youtube_dl/extractor/libraryofcongress.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        self.report_download_webpage(video_id)
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
+        json_id = re.search('media-player-[0-9A-Z]{32}', webpage).group(0).replace('media-player-', '')
+        data = self._parse_json(self._download_webpage('https://media.loc.gov/services/v1/media?id=%s' %json_id, video_id), video_id)
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: There should be spaces between ```%```.

File: youtube_dl/extractor/libraryofcongress.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+import re
+from .common import InfoExtractor
+
+
+class LibraryOfCongressIE(InfoExtractor):
+    IE_NAME = 'LibraryOfCongress'
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: There's no need to include an ```IE_NAME``` which is the same as the extractor name.

File: youtube_dl/extractor/libraryofcongress.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    }, {
+        'url': 'https://www.loc.gov/item/97516576/',
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        self.report_download_webpage(video_id)
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: This function call is already (indirectly) included in ```_download_webpage```.

File: youtube_dl/extractor/libraryofcongress.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
+        json_id = re.search('media-player-[0-9A-Z]{32}', webpage).group(0).replace('media-player-', '')
+        data = self._parse_json(self._download_webpage('https://media.loc.gov/services/v1/media?id=%s' %json_id, video_id), video_id)
+        data = data.get('mediaObject')
+
+        media_url = data.get('derivatives')[0]['derivativeUrl']
+        media_url = media_url.replace('rtmp', 'https')
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: I guess the original RTMP link works, too. Have you tried it?

File: youtube_dl/extractor/libraryofcongress.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+        is_video = data.get('mediaType').lower() == 'v'
+        if not (media_url.endswith('mp4') or media_url.endswith('mp3')):
+            media_url += '.mp4' if is_video else '.mp3'
+
+        if media_url.index('vod/mp4:') > -1:
+            media_url = media_url.replace('vod/mp4:', 'hls-vod/media/') + '.m3u8'
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: Formats with M3U8 playlists should be passed to ```_extract_m3u8_formats``` or manually added the ```protocol``` field.

File: youtube_dl/extractor/libraryofcongress.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        elif url.index('vod/mp3:') > -1:
+            media_url = media_url.replace('vod/mp3:', '')
+
+        if media_url.endswith('.m3u8'):
+            formats = self._extract_m3u8_formats(media_url, video_id, ext='mp4')
+        elif media_url.endswith('.mp3'):
+            formats = media_url
```
Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: ```formats``` is always a list of dictionaries.

File: youtube_dl/extractor/libraryofcongress.py (hunk 11 above)
Comment (TRox1972) [inline youtube_dl/extractor/libraryofcongress.py]: It should work, though when using rtmpdump, it just hangs on 'connecting'.  Would it be better to download the RTMP stream?

File: youtube_dl/extractor/libraryofcongress.py (hunk 6 above)
Comment (TRox1972) [inline youtube_dl/extractor/libraryofcongress.py]: I should use `data = data['mediaObject']` instead, right?

Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: Yes.

File: youtube_dl/extractor/libraryofcongress.py (hunk 5 above)
Comment (TRox1972) [inline youtube_dl/extractor/libraryofcongress.py]: Have changed it to `self._search_regex()`. The thumbnail url can be found as:

`<meta property="og:image" content="http://stream-media.loc.gov/ls/sagan/stills/1882820-6-1_thumb.jpg"/>`
`<meta property="og:image:secure_url" content="https://stream-media.loc.gov/ls/sagan/stills/1882820-6-1_thumb.jpg"/>`
`<meta property="twitter:image" content="https://stream-media.loc.gov/ls/sagan/stills/1882820-6-1_thumb.jpg"/>`

What would a stricter filter look like?

Comment (yan12125) [inline youtube_dl/extractor/libraryofcongress.py]: For these cases, use ```self._og_search_thumbnail``` or ```self._html_search_meta```.

**Comment by TRox1972** [comment]
thanks for feedback @yan12125. Does this look good?

File: youtube_dl/extractor/libraryofcongress.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+            formats.append({
+                'url': media_url,
+                'ext': 'mp3',
+            })
+
+        return {
+            'id': video_id,
+            'thumbnail': self._og_search_thumbnail(webpage),
+            'title': self._og_search_title(webpage),
+            'ext': 'mp4' if is_video else 'mp3',
```
Comment (dstftw) [inline youtube_dl/extractor/libraryofcongress.py]: This is superfluous since you provide `formats`.

File: youtube_dl/extractor/libraryofcongress.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
+        json_id = self._search_regex('media-player-([0-9A-Z]{32})', webpage, 'json id')
+
+        data = self._parse_json(self._download_webpage('https://media.loc.gov/services/v1/media?id=%s' % json_id, video_id), video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/libraryofcongress.py]: Carry long lines.

**Comment by dstftw** [comment]
Cherry-picked in 9c3c447eb389726d98189d972a2d772ef729132d, thanks.
